### ISSUE TITLE
Safecracker: The Ultimate Puzzle Adventure - Compatibility

### ISSUE THREAD

**Issue by master801** [elishacloud/dxwrapper#311]
Reporting compatibility for the game `Safecracker: The Ultimate Puzzle Adventure` (also known as `Safecracker 2`).

OS: Windows 10 Pro 21H2
GPU: Gigabyte AMD RX 580 GAMING 4G (4GB) - AMD Adrenalin `24.9.1`

Game: `Safecracker: The Ultimate Puzzle Adventure` (Steam version)
dxwrapper: `v1.2.7200.25`

Normal behavior (without dxwrapper): Game black screens and flickers some textures immensely.

Fixed behavior (with dxwrapper): Use `d3d9.dll` wrapper, toggle `EnableD3d9Wrapper` to enabled, toggle `ForceDirect3D9On12` in the `[d3d9]` section to enabled. Game no longer black screens and flickers.

**Comment by elishacloud** [comment]
Does this issue happen if you only have `EnableD3d9Wrapper` enabled?

`ForceDirect3D9On12` is not running my code.  I am just calling into [Microsoft's code here](https://github.com/microsoft/D3D9On12).  There are still bugs in `ForceDirect3D9On12`, but you will need to open a case with Microsoft about that.

**Comment by master801** [comment]
Yes, the game will still black screen / flicker if _only_ `EnableD3d9Wrapper` is enabled.

In combination of `EnableD3d9Wrapper` and `ForceDirect3D9On12`, the game will display properly / no longer black screen nor flicker.

**Comment by elishacloud** [comment]
Try enabling `SetSwapEffectShim` rather than `ForceDirect3D9On12`.

By default in Direct3D9 Microsoft sets `SetSwapEffectShim` to 1. However, in dxwrapper I set it to 0 because that has better compatibility with many older games.

**Comment by elishacloud** [comment]
Can you try with the latest build:  [dxwrapper.zip](https://github.com/user-attachments/files/23246885/dxwrapper.zip)

**Comment by master801** [comment]
Hi, with `D3d9on12` enabled, the game works perfectly. No other settings enabled.

With only `EnableD3d9Wrapper` enabled, the game just displays a black screen, even with `SetSwapEffectShim` enabled, as you've suggested.

Some odd (undocumented?) behavior noticed: Enabling `ShowFPSCounter` in the `[d3d9]` section (with `D3d9on12` enabled) will make the FPS counter show up. From this, I assume that this entire section is accessible when `D3d9on12` is toggled?

Side note, but I had to use the `d3d9.dll` from the old build since `ddraw.dll` only came in that zip (for some reason while the game's exe references `ddraw.dll`, it doesn't seem to load it at all from the path.)

Renaming `ddraw.dll` to `d3d9.dll` also worked - which I assume is because it's just a DLL hooker.

**Comment by elishacloud** [comment]
> Hi, with `D3d9on12` enabled, the game works perfectly. No other settings enabled.

Ok, very good.

> With only `EnableD3d9Wrapper` enabled, the game just displays a black screen, even with `SetSwapEffectShim` enabled, as you've suggested.

Sounds like an issue with Direct3D9 and not dxwrapper.  dxwrapper's Driect3D9 support is mainly just a pass-through stub with options for enabling a few features.

> Enabling `ShowFPSCounter` in the `[d3d9]` section (with `D3d9on12` enabled) will make the FPS counter show up.

Yes.  The game is a d3d9 game.  Virtually all settings under the `[d3d9]` section will work with `D3d9on12` enabled.

> Renaming `ddraw.dll` to `d3d9.dll` also worked - which I assume is because it's just a DLL hooker.

Yes.  These are just stub loaders.  They are all the same.  You can rename them to any supported dll stub name.

**Comment by elishacloud** [comment]
According to this wiki: https://www.pcgamingwiki.com/wiki/Safecracker:_The_Ultimate_Puzzle_Adventure

It looks like the game uses Direct3D9.  Try using the d3d9.dll stub and just enable `EnableD3d9Wrapper` only.  You can try with either `D3d9to9Ex` or `D3d9on12` also.

Please make sure you are using the latest release.

_Note: `EnableD3d9Wrapper` isn't intended to fix any game issue.  It is just there to allow adding features into games, like anti-aliasing._
